File: backend/app/schemas/processing_job.py
"""Schemas for `app.models.processing_job.ProcessingJob`."""
import uuid
from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, Field

from app.models.enums import ProcessingJobStatus, ProcessingJobType
from app.schemas.common import ORMModel, PaginatedResponse, TimestampSchema


class ProcessingJobBase(BaseModel):
    job_type: ProcessingJobType


class ProcessingJobCreate(ProcessingJobBase):
    document_version_id: uuid.UUID


class ProcessingJobUpdate(BaseModel):
    """Covers the fields a worker sets as a job progresses:
    pending -> running -> completed/failed.
    """

    status: ProcessingJobStatus | None = None
    error_message: str | None = None
    retry_count: Annotated[int, Field(ge=0)] | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None


class ProcessingJobResponse(ProcessingJobBase, TimestampSchema, ORMModel):
    id: uuid.UUID
    document_version_id: uuid.UUID
    status: ProcessingJobStatus
    error_message: str | None = None
    retry_count: Annotated[int, Field(ge=0)]
    started_at: datetime | None = None
    completed_at: datetime | None = None


ProcessingJobListResponse = PaginatedResponse[ProcessingJobResponse]

File: frontend/src/types/processing-job.ts
export type ProcessingJobType = 'upload' | 'parse' | 'chunk' | 'embed' | 'index'

export type ProcessingJobStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'

export interface ProcessingJob {
  id: string
  document_version_id: string
  job_type: ProcessingJobType
  status: ProcessingJobStatus
  error_message: string | null
  retry_count: number
  started_at: string | null
  completed_at: string | null
  created_at: string
  updated_at: string
}

export type ProcessingJobListResponse = import('./api').PaginatedResponse<ProcessingJob>
